### PR TITLE
Improve character selection.

### PR DESCRIPTION
**Pull request by zaknelson**

### CODE REVIEW

File: Classes/characters/CharacterSystem.cpp [hunk 1]
```diff
@@ -52,6 +80,7 @@ void CharacterSystem::addListeners() {
     listener_ = EventListenerTouchOneByOne::create();
     listener_->onTouchBegan = [&](Touch* touch, Event* event) {
         touchBeginLocation_ = touch->getLocation();
+        characterTouchedAtTouchBegin_ = getCharacterAtPoint(touchBeginLocation_);
```
Comment (zaknelson) [inline Classes/characters/CharacterSystem.cpp]: The Blizzard selection model.

File: Classes/items/SlingshotComponent.cpp [hunk 2]
```diff
@@ -9,7 +9,9 @@ SlingshotComponent::SlingshotComponent() : collisionTime_(std::numeric_limits<fl
 
 SlingshotComponent::~SlingshotComponent() {
     if (drawNode_) {
+        drawNode_->clear();
```
Comment (zaknelson) [inline Classes/items/SlingshotComponent.cpp]: Unrelated but wanted to keep it similar to `ActiveCharacterComponent`.

File: Classes/characters/CharacterSystem.cpp [hunk 3]
```diff
@@ -25,6 +25,34 @@ CharacterSystem::~CharacterSystem() {
...
+        Entity* character = activeCharacters[i];
+        ActiveCharacterComponent* activeCharacterComponent = dynamic_cast<ActiveCharacterComponent*>(character->getComponent(ComponentType::ACTIVE_CHARACTER));
+        DrawNode* drawNode = activeCharacterComponent->getDrawNode();
+        if (!drawNode) {
+            drawNode = DrawNode::create();
+            activeCharacterComponent->setDrawNode(drawNode);
+            uiLayer_->addChild(drawNode);
+        }
+        drawNode->clear();
+        Vec2 origin = CharacterUtils::getTop(character);
```
Comment (maxnelso) [inline Classes/characters/CharacterSystem.cpp]: "origin" seems like a strange name for this variable, isn't it just the top?

File: Classes/characters/CharacterSystem.cpp [hunk 4]
```diff
@@ -116,42 +145,28 @@ void CharacterSystem::addListeners() {
             }
         }
     });
-
 }
 
-void CharacterSystem::inputInPolygon(std::vector<Vec2>& polygon) {
+EntityID CharacterSystem::getCharacterAtPoint(const Vec2& point) {
```
Comment (maxnelso) [inline Classes/characters/CharacterSystem.cpp]: This also seems like a weird name, what about "inputInPolygon"? jk

File: Classes/characters/CharacterSystem.cpp [hunk 5]
```diff
@@ -116,42 +145,28 @@ void CharacterSystem::addListeners() {
...
-            found = true;
-            entityManager_->updateEntityWithComponentsToAdd(characters[i], {new SimpleFlagComponent(ComponentType::ACTIVE_CHARACTER)});
-        } else if (characters[i]->hasComponents(activeCharacterMask)) {
-            entityManager_->updateEntityWithComponentsToRemove(characters[i], {characters[i]->getComponent(ComponentType::ACTIVE_CHARACTER)});
+        float distanceToPoint = center.distance(point);
+        if (distanceToPoint < distanceToNearestCharacter) {
+            distanceToNearestCharacter = distanceToPoint;
+            Size size = ellipseComponent->getSize();
+            size = size * transformComponent->getScale();
+            if (MathUtils::doesEllipseContainsPoint(center, size, point)) {
```
Comment (maxnelso) [inline Classes/characters/CharacterSystem.cpp]: For another commit maybe, but this should certainly be called MathUtils::doesEllipseContainPoint

**Comment by maxnelso** [comment]
Other than that, looks good to me

File: Classes/characters/CharacterSystem.cpp (hunk 3 above)
Comment (zaknelson) [inline Classes/characters/CharacterSystem.cpp]: Agreed. Will clean.

File: Classes/characters/CharacterSystem.cpp (hunk 5 above)
Comment (zaknelson) [inline Classes/characters/CharacterSystem.cpp]: Nah, I-I'll just beat him now.

**Comment by zaknelson** [comment]
@maxnelso Fixed!

**Pull request merged**